### PR TITLE
Support for importing stored data from libolm

### PR DESCRIPTION
**Pull request by richvdh**
Fixes https://github.com/matrix-org/matrix-rust-sdk-crypto-wasm/issues/78

~~Based on https://github.com/matrix-org/matrix-rust-sdk-crypto-wasm/pull/76~~

### CODE REVIEW

File: src/libolm_migration.rs
Comment (bnjbvr) [inline src/libolm_migration.rs:406]: `.context` here too

File: tests/libolm_migration.test.ts
Comment (bnjbvr) [inline tests/libolm_migration.test.ts:77]: Is this something you wanted to address before landing this?

**Review by bnjbvr — CHANGES_REQUESTED** [review]
Sweet! Mostly rust-isms in my comments, it would be super nice if someone from the crypto team could double-check the crypto logic here.

File: src/libolm_migration.rs
Comment (richvdh) [reply src/libolm_migration.rs]: I don't think it's helpful to dive into the gory details of the format (it's an encrypted blob containing a binary structure which is nominally a libolm implementation detail), but I've updated the comment to hopefully better explain where it comes from.

File: src/libolm_migration.rs
Comment (richvdh) [reply src/libolm_migration.rs]: Nice tip about the `Option` extension, thanks!

> Alternatively, since the user_id is mandatory in the BaseMigrationData, let's make it a non Optional field?

Well... that would mean we'd have to pass in a UserID in the constructor of `BaseMigrationData`, and once we go down that path, pretty soon we get to a situation where the constructor takes 150 parameters. The whole point of `BaseMigrationData` is to avoid functions that take lots of parameters and instead allow the javascript side to use named properties.

File: src/libolm_migration.rs
Comment (richvdh) [reply src/libolm_migration.rs]: Well, first of all: it only matters if we have run out of one-time keys (OTKs) on the server; in which case, yes, until we receive the first `/sync` response, we won't upload any new keys. But it's worth noting that the fact we're starting our application doesn't make much difference here: most likely, we ran out of OTKs before the app even started. Waiting a few more minutes until we get the first `/sync` response won't make much difference here.

Secondly: the matrix protocol handles the "out of OTKs" situation by having a "fallback key", which is used as a last resort once OTKs have run out (and is replaced once the application starts and discovers the fallback key has been used). It has slightly inferior security properties to a proper OTK, but means that running out of OTKs isn't a disaster.

File: src/libolm_migration.rs
Comment (richvdh) [reply src/libolm_migration.rs]: again, have updated the comment.

File: src/libolm_migration.rs
Comment (richvdh) [reply src/libolm_migration.rs:244]: I might be missing something, but `Date` doesn't implement `Default`, so `derive(Default)` doesn't work.

File: src/libolm_migration.rs
Comment (richvdh) [reply src/libolm_migration.rs]: noted. updated.

File: src/libolm_migration.rs
Comment (richvdh) [reply src/libolm_migration.rs]: as before: updated the comment.

File: src/libolm_migration.rs
Comment (richvdh) [reply src/libolm_migration.rs]: this method returns `Result<_, JsError>`, rather than `std::Error` or whatever, so I don't think `.context` works here. Have switched to `.ok_or_else` though.

File: tests/libolm_migration.test.ts
Comment (richvdh) [reply tests/libolm_migration.test.ts:77]: Not really, tbh. It looks *tricky*.

**Comment by richvdh** [comment]
@bnjbvr: Thanks for the review! Please take another look.

File: src/libolm_migration.rs
Comment (bnjbvr) [reply src/libolm_migration.rs:161]: (Yes I meant that, modulo order of words :sweat_smile:) Thanks for the explanations!

File: src/libolm_migration.rs
Comment (bnjbvr) [inline src/libolm_migration.rs:234]: Just seeing this: it might be possible that `pickle_key.to_vec()` could be hoisted above the `let rust_sessions = ` statement, to avoid creating the pickle key vector for every single session.

File: src/libolm_migration.rs
Comment (bnjbvr) [inline src/libolm_migration.rs:374]: Ditto here

File: src/libolm_migration.rs
Comment (bnjbvr) [reply src/libolm_migration.rs]: I think there's a middle ground where the two required parameters are provided (user id and device id), and then every other optional parameter can be set manually. But I won't insist since this is bindings code, as long as you're happy with the API shape here.

**Review by bnjbvr — APPROVED** [review]
lgtm

File: src/libolm_migration.rs
Comment (richvdh) [reply src/libolm_migration.rs]: Worth noting that `pickled_account` is also required: it's just that we don't have an explicit check for it here because it's not strongly typed (it defaults to an empty string, which is an invalid value), so strictly speaking that's three required parameters.

I think the current arrangement is the least bad solution in terms of the js-side API.

File: src/libolm_migration.rs
Comment (BillCarsonFr) [inline src/libolm_migration.rs:79]: Maybe worth addind in the name of the field (or in the binding name) that it's the private part? Same for the 4 secrets?

File: src/libolm_migration.rs
Comment (BillCarsonFr) [inline src/libolm_migration.rs:161]: Maybe we should `.zeroize()` the `data.cross_signing_*` at this point?

File: src/libolm_migration.rs
Comment (BillCarsonFr) [inline src/libolm_migration.rs:376]: If one of the megolm session to import is for some reason malformed/corrupted, will the whole chunk fail to import?
Maybe it would be better to ignore "malformed" and log something but not interrupt all?

File: src/libolm_migration.rs
Comment (BillCarsonFr) [inline src/libolm_migration.rs:369]: This could be potentially a lot of sessions. Should we inform the caller to do that in chunks? Or have some progress reporting?

File: src/libolm_migration.rs
Comment (BillCarsonFr) [inline src/libolm_migration.rs:420]: I wonder what `history_visibility` is used for InboundGroupSession. Is it for future support of shared history on invite?

File: src/libolm_migration.rs
Comment (BillCarsonFr) [inline src/libolm_migration.rs:56]: Shouldn't there be a list of tracked users here? or a `migrate_tracked_users` method?

**Review by BillCarsonFr — CHANGES_REQUESTED** [review]
Just a few comments. I think that the migration of tracked users is missing?

File: src/libolm_migration.rs
Comment (richvdh) [reply src/libolm_migration.rs:376]: I'm not entirely convinced. This sounds like we could end up not migrating *any* sessions, and the user would just be confused.

Also: once the batch is successfully imported, we *delete* it from the legacy store. So, if we do fail to import any sessions, we'd end up completely throwing them away, which seems very dangerous.

I'd prefer to leave it and see if any errors actually arise in practice.

File: src/libolm_migration.rs
Comment (richvdh) [reply src/libolm_migration.rs:369]: The intention is that the application calls this with small enough batches that there is no need for an intermediate progress update.

File: src/libolm_migration.rs
Comment (BillCarsonFr) [reply src/libolm_migration.rs:376]: Not exacty related but we have seen sort of similar things when [importing from backup](https://github.com/element-hq/element-web/issues/26774), also when migrating tracked users (legacy is more lax on userIds).

Could we return a result with the succesfully migrated things? so that the caller would delete only these and handle the error on the application side?
I'll let you decide what's best

File: src/libolm_migration.rs
Comment (BillCarsonFr) [reply src/libolm_migration.rs:369]: Maybe document that?

File: src/libolm_migration.rs
Comment (richvdh) [reply src/libolm_migration.rs:420]: A very good question. I couldn't find anywhere that it was used.

It's initialised to `None` in lots of other places, so hopefully it's safe.

**Review by BillCarsonFr — APPROVED** [review]

File: src/libolm_migration.rs
Comment (richvdh) [reply src/libolm_migration.rs:161]: I wondered about that. I concluded it's not worthwhile, because even if we zeroize it on the Rust side, there may well still be copies kicking around in memory from the javascript side which we have no hope of zeroizing. (As https://github.com/veorq/cryptocoding#clean-memory-of-secret-data says: "there's virtually no way to reliably clean secret data in garbage-collected languages (such as ... Javascript)").

And we don't generally zeroize any other sensitive data that is passed through the wasm bindings.

File: src/libolm_migration.rs
Comment (richvdh) [reply src/libolm_migration.rs:376]: > Could we return a result with the succesfully migrated things? so that the caller would delete only these and handle the error on the application side?

Well, if we don't delete the broken sessions, then we'll just get stuck in a loop of repeatedly trying to migrate them. (There is no record of "which sessions have been migrated").)

I suppose one thing that we might think about is whether we can display an error to the user if migration goes wrong.

File: src/libolm_migration.rs
Comment (richvdh) [reply src/libolm_migration.rs:56]: As discussed elsewhere: I don't think there's any need to migrate the tracked users.

File: src/libolm_migration.rs
Comment (richvdh) [reply src/libolm_migration.rs:378]: 808a5736

**Pull request merged**
